File: day14/day14.cpp
#include <fstream>
#include <numeric>
#include <regex>
#include <unordered_map>
#include <vector>
#include <fmt/format.h>

using namespace std;

static const regex MASKP("mask = ([01X]{36})");
static const regex MEMP("mem\\[(\\d+)\\] = (\\d+)");

static size_t part1(const vector<string>& program)
{
	unordered_map<size_t, size_t> mem;
	string mask;
	for (auto& line: program)
	{
		smatch sm;
		if (regex_match(line, sm, MASKP))
		{
			mask = sm[1];
		}
		else if (regex_match(line, sm, MEMP))
		{
			size_t address = stoull(sm[1]);
			size_t before = stoull(sm[2]);
			size_t after = 0;
			for (size_t i = 0; i < 36; i++)
			{
				switch (mask[i])
				{
				case 'X': after |= before & (1ULL<<(35-i)); break;
				case '1': after |= 1ULL<<(35-i); break;
				case '0': break;
				}
			}
			mem[address] = after;
		}
	}
	return accumulate(
		mem.cbegin(),
		mem.cend(),
		0ULL,
		[](auto acc, const auto& v) {
			return acc + v.second;
		});
}

static void gen(const string& pattern, size_t pos, string& cur, vector<size_t>& out)
{
	if (pos == pattern.size())
	{
		size_t val = 0;
		for (auto c: cur)
		{
			val |= c == '1';
			val <<= 1;
		}
		out.push_back(val);
	}
	else if (pattern[pos] != 'X')
	{
		cur[pos] = pattern[pos];
		gen(pattern, pos+1, cur, out);
	}
	else
	{
		cur[pos] = '0';
		gen(pattern, pos+1, cur, out);
		cur[pos] = '1';
		gen(pattern, pos+1, cur, out);
	}
}

static size_t part2(const vector<string>& program)
{
	unordered_map<size_t, size_t> mem;
	vector<size_t> addresses;
	string mask;
	for(auto& line: program)
	{
		smatch sm;
		if (regex_match(line, sm, MASKP))
		{
			mask = sm[1];
		}
		else if (regex_match(line, sm, MEMP))
		{
			size_t addr = stoull(sm[1]);
			size_t result = stoull(sm[2]);
			string pattern(36, ' ');
			for (size_t i = 0; i < 36; i++)
			{
				switch(mask[i])
				{
				case 'X':
				case '1':
					pattern[i] = mask[i];
					break;
				case '0':
					if (addr & (1ULL << 35))
					{
						pattern[i] = '1';
					}
					else
					{
						pattern[i] = '0';
					}
					break;
				}
				addr <<= 1;
			}
			addresses.clear();
			string tmp('0', 36);
			gen(pattern, 0, tmp, addresses);
			for (auto a: addresses)
			{
				mem[a] = result;
			}
		}
	}
	return accumulate(
		mem.cbegin(),
		mem.cend(),
		0ULL,
		[](auto acc, const auto& v) {
			return acc + v.second;
		});
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<string> program;
	string line;
	while (getline(input, line))
	{
		program.emplace_back(move(line));
	}
	input.close();

	fmt::print("Part1: {}\nPart2: {}\n", part1(program), part2(program));
	return 0;
}

File: day14/day14.py
#!/usr/bin/env python3

import re
import sys

MASKP = re.compile(r"mask = ([X10]{36})")
MEMP = re.compile(r"mem\[(\d+)\] = (\d+)")

def part1(program):
    mem = {}
    for line in program.strip().split("\n"):
        m = MASKP.match(line)
        if m:
            mask = list(m[1])
            continue

        m = MEMP.match(line)
        if m:
            address = int(m[1])
            source = int(m[2])
            result = 0
            for i, symbol in enumerate(mask):
                if symbol == "X":
                    result |= source & (1<<(35-i))
                elif symbol == "1":
                    result |= 1<<(35-i)
                else:
                    pass

            mem[address] = result

    return sum(mem.values())

def gen(pattern, i, cur, lst):
    if i == len(pattern):
        num = 0
        for digit in cur:
            num |= (digit == "1") and 1 or 0
            num <<= 1
        lst.append(num)
    elif pattern[i] != "X":
        gen(pattern, i+1, cur + pattern[i], lst)
    else:
        gen(pattern, i+1, cur + "0", lst)
        gen(pattern, i+1, cur + "1", lst)

def part2(program):
    mem = {}
    for line in program.strip().split("\n"):
        m = MASKP.match(line)
        if m:
            mask = list(m[1])
            continue

        m = MEMP.match(line)
        if m:
            address = int(m[1])
            result = int(m[2])
            pattern = []
            for i, symbol in enumerate(mask):
                if symbol != "0":
                    pattern.append(symbol)
                elif address & (1<<(35-i)) != 0:
                    pattern.append("1")
                else:
                    pattern.append("0")

            addrlist = []
            gen(pattern, 0, "", addrlist)
            for addr in addrlist:
                mem[addr] = result

    return sum(mem.values())

# txt = """mask = XXXXXXXXXXXXXXXXXXXXXXXXXXXXX1XXXX0X
# mem[8] = 11
# mem[7] = 101
# mem[8] = 0"""
# assert part1(txt) == 165

# txt = """mask = 000000000000000000000000000000X1001X
# mem[42] = 100
# mask = 00000000000000000000000000000000X0XX
# mem[26] = 1"""
# assert part2(txt) == 208

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", part1(txt))
    print("Part2:", part2(txt))
